### DISCUSSION TITLE
Suggested plugin hook: before/after rendering

### DISCUSSION THREAD

**Discussion by ThrawnCA** [ckan/ckan#6036] [Ideas]
It would sometimes be useful for plugins to be able to hook into the global page rendering process (in `base.render`), either to update the template variables beforehand, or to perform extra processing on the result. One specific use case is injecting CSRF tokens into all POST forms.

Monkey-patching `base.render` doesn't work, because `ckan.plugins.toolkit` may already have taken a copy of the original function and distributed it to other extensions. Monkey-patching both `base.render_jinja2' and `base.flask_render_template` does work, but is inelegant and prone to breakage in future CKAN versions.

**Comment by ThrawnCA** [comment]
FYI The Flask `after_app_request` hook is partially usable, but it appears that `g.userobj` is no longer available at that point, which makes our CSRF filtering harder (we use HMAC-based tokens with the username being one of the inputs).

**Reply by smotornyuk** [reply]
There is #5359 PR under review that supplies codebase with signals. If this type of hook is sufficient, we can just put the required signals in the most suitable places. Can you take a look at this PR  and if you have any suggestions, i can add signals with any information you think is usable

**Reply by ThrawnCA** [reply]
Interesting idea, but signals won't do the job we want here. Point 4 says to keep using interfaces when you want to modify parameters or return a different value, which are the intended use cases for this hook. Specifically, we want to alter the HTML response body just before it's sent to the client.

**Reply by ThrawnCA** [reply]
We've gone ahead with using Flask `after_app_request`, but it's less than ideal, because we can only access the response body by disabling streaming mode, which likely has a performance impact. A hook at the point of actually rendering the page would work better.
